dkist/net/helpers.py

```diff
@@ -0,0 +1,95 @@
+"""
+from os import PathLike
+from typing import Union, Literal
+from pathlib import Path
+
+from sunpy.net.base_client import QueryResponseRow
+
+from dkist.net.attrs import Dataset
+from dkist.net.client import DKISTClient, DKISTQueryResponseTable
+from dkist.net.globus.transfer import _orchestrate_transfer_task
+
+__all__ = ['transfer_whole_dataset']
+
+
+def _get_dataset_inventory(dataset_id: str):  # pragma: no cover
+    """
+    Do a search for a single dataset id
+    """
+    return DKISTClient().search(Dataset(dataset_id))
+
+
+def transfer_whole_dataset(dataset: Union[str, QueryResponseRow, DKISTQueryResponseTable],
+                           path: PathLike = "/~/",
+                           destination_endpoint: str = None,
+                           progress: Union[bool, Literal["verbose"]] = True,
+                           wait: bool = True):
+    """
+    Transfer a whole dataset to a path on a globus endpoint.
+
+    Parameters
+    ----------
+    dataset
+        The dataset to transfer. This can either be a string dataset id, or it
+        can be a single row from a `.DKISTQueryResponseTable` as returned in a
+        call to ``Fido.search``.
+
+    destination_endpoint
+        A unique specifier for a Globus endpoint. If `None` a local
+        endpoint will be used if it can be found, otherwise an error will
+        be raised. See `~dkist.net.globus.get_endpoint_id` for valid
+        endpoint specifiers.
+
+    progress
+        If `True` status information and a progress bar will be displayed
+        while waiting for the transfer to complete.
+        If ``progress="verbose"`` then all globus events generated during the
+        transfer will be shown (by default only error messages are shown.)
+
+    wait
+       If `False` then the function will return while the Globus transfer task
+       is still running. Setting ``wait=False`` implies ``progress=False``.
+
+    Returns
+    -------
+    destination_path
+        The path to the directory containing the dataset on the destination
+        endpoint.
+
+    """
+    # Avoid circular import
+    from dkist.net import conf
+
+    if isinstance(dataset, str):
+        dataset = _get_dataset_inventory(dataset)[0]
+
+    if isinstance(dataset, DKISTQueryResponseTable) and len(dataset) > 1:
+        for record in dataset:
+            transfer_whole_dataset(record,
+                                   path=path,
+                                   destination_endpoint=destination_endpoint,
+                                   progress=progress,
+                                   wait=wait)
+
+    dataset_id = dataset["Dataset ID"]
+    proposal_id = dataset["Primary Proposal ID"]
+    bucket = dataset["Storage Bucket"]
+
+    destination_path = Path(path) / proposal_id
+
+    file_list = [Path(conf.dataset_path.format(
+        datasetId=dataset_id,
+        primaryProposalId=proposal_id,
+        bucket=bucket
```

Comment:
This usage constrains the pattern in `conf.dataset_path` and so if that path ever changes you'll need to update these lines as well. Is there a way to keep all of that logic in a single place?

Reply:
Yeah, I know, but I can't think of a way out of it shy of a large refactoring of the Fido client.